### PR TITLE
Catch unknown method during upgrade repair step

### PR DESCRIPTION
**Pull request by ChristophWurst**
It's our friend the upgrade error.

How it blows up
* The Nextcloud process doing the upgrade loads all application code. Mail app code as well
* The migration is done
* Post migration steps are run
  + The new repair step is loaded, but it uses a class that existed before. That class isn't loaded another time. Kaboom.

The same trick is also found in `\OCA\Mail\Migration\AddMissingMessageIds::run`.

### CODE REVIEW

File: lib/Migration/RepairMailTheads.php [hunk 1]
```diff
@@ -47,6 +48,16 @@ public function getName(): string {
 	}
 
 	public function run(IOutput $output): void {
+		/**
+		 * During the upgrade to v1.11.3 and later the old version of the
+		 * mapper is loaded before this new repair step is performed. Hence even after
+		 * the program code got replaced, the class doesn't have the new method. We
+		 */
+		if (!method_exists($this->mapper, 'resetInReplyTo')) {
+			$output->warning('New Mail code hasn\'t been loaded yes, skipping message ID repair. Please run `occ maintenance:repair` after the upgrade.');
```
Comment (nickvergessen) [inline lib/Migration/RepairMailTheads.php:57]:
```suggestion
			$output->warning('New Mail code hasn\'t been loaded yet, skipping message ID repair. Please run `occ maintenance:repair` after the upgrade.');
```

Comment (ChristophWurst) [reply lib/Migration/RepairMailTheads.php:57]: yet you are right

**Review by miaulalala — APPROVED** [review]

File: lib/Migration/RepairMailTheads.php [hunk 2]
```diff
@@ -47,6 +48,16 @@ public function getName(): string {
...
+		 * During the upgrade to v1.11.3 and later the old version of the
+		 * mapper is loaded before this new repair step is performed. Hence even after
+		 * the program code got replaced, the class doesn't have the new method. We
+		 */
+		if (!method_exists($this->mapper, 'resetInReplyTo')) {
+			$output->warning('New Mail code hasn\'t been loaded yet, skipping message ID repair. Please run `occ maintenance:repair` after the upgrade.');
+			return;
+		}
+
 		$count = $this->mapper->resetInReplyTo();
```
Comment (raimund-schluessler) [inline lib/Migration/RepairMailTheads.php:61]: @ChristophWurst @miaulalala Sorry, I am late here. But couldn't this method be located directly in the migration step, e.g. in the `postSchemaChange`? The method is only used in the repair step and you wouldn't need to run `occ maintenance:repair` afterwards manually.

Comment (ChristophWurst) [reply lib/Migration/RepairMailTheads.php:61]: that would be possible as a workaround.

personally I would rather push for finally fixing repair steps than creating even more workaround everywhere

**Pull request merged**